### PR TITLE
(CFACT-227) Move locale and catch to leatherman

### PR DESCRIPTION
**Pull request by branan**
These are a bit lighter-weight than the logging library, so are a
quicker start for getting it all put together.

This can be tested against https://github.com/branan/cfacter/tree/imp/master/CFACT-227-leatherman

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+# This macro gives us a pile of common variable names and behaviors
+# for leatherman libraries that we can work with
+#
+# * all the common LEATHERMAN_<LIBNAME>_{LIB,LIBS,INCLUDE,DEPS}
+#   variables for integrating individual leatherman libraries into a
+#   top-level project
+# * automatic test integration with the top-level test executable
+# * variables passed into each subdir for common naming conventions
+# * options to enable/disable individual libraries
+macro(add_leatherman_dir dir)
```
Comment (MikaelSmith) [inline CMakeLists.txt]: Can we pull everything about add_leatherman_dir into a separate CMake module?

File: catch/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# Just point the user at our vendored catch. Nothing else to do here
```
Comment (MikaelSmith) [inline catch/CMakeLists.txt]: What's the value in this redirect?

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    set(${lib_var} "${libname}" PARENT_SCOPE)
+endmacro()
+
+macro(add_leatherman_test)
+    add_library(${testlibname} STATIC ${ARGV})
+    set(${testlib_var} "${testlibname}" PARENT_SCOPE)
+endmacro()
+
+# This macro gives us a pile of common variable names and behaviors
+# for leatherman libraries that we can work with
```
Comment (MikaelSmith) [inline CMakeLists.txt]: The comments should point out the pattern used for enabling libraries.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+add_leatherman_dir(catch EXCLUDE_FROM_VARS)
+add_leatherman_dir(locale)
+
+export_var(LEATHERMAN_INCLUDE_DIRS)
+export_var(LEATHERMAN_LIBRARIES)
+
+if(LEATHERMAN_ENABLE_TESTING)
+    enable_testing()
+    add_subdirectory(tests)
```
Comment (MikaelSmith) [inline CMakeLists.txt]: I kind of like having tests in each individual library, so it's obvious which libraries are covered.

**Comment by MikaelSmith** [comment]
I don't know if specifying install files makes more sense in this project or in the projects that use it. I lean towards leaving it out here until we see a reason to change it.

**Comment by MikaelSmith** [comment]
We should add a README on how to use and include libraries.

File: catch/CMakeLists.txt (hunk 2 above)
Comment (branan) [inline catch/CMakeLists.txt]: Avoiding a special-case around catch in the top-level CMake file

File: CMakeLists.txt (hunk 4 above)
Comment (branan) [inline CMakeLists.txt]: That's the intended pattern. This is just for the test executable. I'll rename the directory and clarify that with a comment down here.

File: CMakeLists.txt (hunk 1 above)
Comment (branan) [inline CMakeLists.txt]: definitely

File: catch/CMakeLists.txt (hunk 2 above)
Comment (MikaelSmith) [inline catch/CMakeLists.txt]: Alright, I wonder about having the actual source over in vendor/catch then. We could point out that it's vendored in some other way.

Comment (MikaelSmith) [inline catch/CMakeLists.txt]: I see why this makes sense (we may eventually have vendored libraries that we don't plan to export), and the vendor dir pattern is a useful identifier. So this is fine.

File: CMakeLists.txt (hunk 4 above)
Comment (MikaelSmith) [inline CMakeLists.txt]: Oh I see, each library adds a test lib.

**Comment by branan** [comment]
@MikaelSmith it's probably worth having a way to build this standalone (as a pile of static libraries and headers), but I don't see it as needed for the initial conversion.

README is in-progress.

**Comment by branan** [comment]
@MikaelSmith big update!

* Added README with usage instructions as well as a basic primer on adding a library
* Moved all the macros into their own `.cmake` file
* Cleaned up macro documentation

File: README.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+* `LEATHERMAN_INCLUDE_DIRS` The include paths of all enabled
+  leatherman libraries
+* `LEATHERMAN_LIBRARIES` The library names of all enabled leatherman
+  libraries, as well as their dependencies.
+
+In addition, each enabled library sets a number of library-specific
+variables:
+
+* `LEATHERMAN_<LIBRARY>_INCLUDE` The include directoryor directories
```
Comment (MikaelSmith) [inline README.md]: Need a space in `directoryor`.

**Comment by MikaelSmith** [comment]
This looks great. :+1:

**Comment by MikaelSmith** [comment]
Is this still WIP? Looks ready to merge to me.

File: cmake/leatherman.cmake [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	    debug("Excluding values for ${id_upper} from common vars")
+	endif()
+    endif()
+endmacro(add_leatherman_dir)
+
+####
+# Helper Macros
+####
+
+# Usage: debug("Somethign cool is happening")
```
Comment (kylog) [inline cmake/leatherman.cmake]: 'Something'

**Comment by kylog** [comment]
:+1: 

Great work!

**Comment by kylog** [comment]
@parisiale @ploubser @richardc thoughts?

**Comment by ploubser** [comment]
Sorry, commented on #2 This looks really cool. :+1:

**Comment by richardc** [comment]
Seems OK, looking forward to the boost-style `Include(leatherman logging catch)` style of declaring which modules are wanted.

**Comment by parisiale** [comment]
Leatherman looks great.
If we agree, I can make a PR for adding the thread ID and source line to the logging format; here's how it would look like: http://git.io/AFV6.

**Comment by branan** [comment]
@MikaelSmith @kylog Typos addressed. I think this is ready to be merged now

**Comment by dvhwgumby** [comment]
Sorry to add to a four year old open issue, but I cannot use the Leatherman::locale component in CMake.  `find_package(Leatherman REQUIRED locale util)` works, and I can build my code by putting `${LEATHERMAN_LIBRARIES}` in the appropriate `target_link_libraries` entries.

But if I try to do things the "correct" way (by putting `Leatherman::locale` in `target_link_libraries`, CMake complains that the target was not found.  My CMake-fu is inadequate to debug this.

I pulled on this thread while beginning to work on a pull request for cpp-hocon issue 103 (not sure how to link references across github projects).

**Comment by MikaelSmith** [comment]
I guess I'm not sure why you say putting `Leatherman::local` in `target_link_libraries` is the correct way? With CMake, you can add something like `message("Libs: ${LEATHERMAN_LIBRARIES}` after `include(leatherman)` to see how the variable's initialized. For cpp-hocon it appears to be `leatherman_util;leatherman_locale;/usr/local/lib/libboost_locale-mt.dylib;/usr/local/lib/libboost_system-mt.dylib;/usr/local/lib/libboost_date_time-mt.dylib;/usr/local/lib/libboost_chrono-mt.dylib;leatherman_util`. So try adding `leatherman_locale`; you'll also need to manually add any dependencies like boost locale.

**Comment by dvhwgumby** [comment]
Indeed, manipulating the various includes are what I do now for Leatherman and hocon-cpp.

However CMake doesn't want users mucking with those files, especially since (who knows?) perhaps some build systems don't use them (all the ones **I** use do, but still). 

target libraries & module names are supposed to manage both includes and library files (.a, .dylib, .so, whatever).  They want us to let CMake do the heavy lifting.  For a long time I thought the poorl named `target_link_libraries` was for `-l` and `-L` but really it is "these are the third party libraries depended on -- make sure cpp can find the headers and ld can find the library objects"

Thus the preferred, and simpler mode is the user doing the merely following:

```
foo.cc:
  #include <boost/locale>
  ....

*CMakeLists.txt:*
  find_package(Boost REQUIRED COMPONENTS locale)

  add_executable(foo foo.cc)
  target_link_libraries (foo Boost::locale)
```

And if I used Leatherman locale:

```
foo.cc:
  #include <leatherman/locale>
  ....

CMakeLists.txt:

  find_package(Leatherman REQUIRED COMPONENTS locale)

  add_executable(foo foo.cc)
  target_link_libraries (foo Leatherman::locale)
```

And let Leatherman manage its own need for Boost etc transparently to the user.  

This is what I wanted to do in a pull request for HOCON but I couldn't get Leatherman to DTRT.  There's enough going on in the Leatherman .cmake files that I'm afraid to break them.

Eventually I will want to compile with VS, which I know nothing about, so I'm eager to be able to tell CMake everything I can at as high a level as I can.

**Comment by MikaelSmith** [comment]
That seems like a pattern that may have been undocumented when we wrote the Leatherman CMake files (circa CMake 3.2). So I'm not familiar with it, and don't know off the top of my head why Leatherman wouldn't support it. It's probable Leatherman needs some work to let CMake manage inherited dependencies on targets.

**Comment by dvhwgumby** [comment]
Yeah, Cmake's documentation, while surprisingly comprehensive, still manages to be horrible.

I don't mean to encourage gratuitous meta work on your company's CMake infrastructure.  I'm grateful you guys ship module support in lib/cmake/leatherman at all.

If you guys ever do get around to updating the leatherman CMake module definition I'll do that for hocon and send a PR.

**Pull request merged**